### PR TITLE
Hug JSX text after multi line tags

### PR DESCRIPTION
**Pull request by karl**
A fix for https://github.com/prettier/prettier/issues/1581

This PR no longer allows a line break between tags and text in JSX. This fixes an issue that was blocking adoption at Facebook (https://github.com/prettier/prettier/issues/1581#issuecomment-300975032) with little adverse effects for our standard formatting.

### CODE REVIEW

File: tests/jsx-text-wrap/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -188,8 +197,9 @@ function DiffOverview(props) {
       <div className="alert alert-info">
         <p>
           This diff overview is computed against the current list of records in
-          this collection and the list it contained on <b>{humanDate(since)}</b>
-          .
+          this collection and the list it contained on <b>
+            {humanDate(since)}
+          </b>.
```
Comment (karl) [inline tests/jsx-text-wrap/__snapshots__/jsfmt.spec.js.snap]: This looks like an edge case with the current implementation. When we have tag that would just fit on a line and it is followed by text it causes this unusual formatting.

I imagine in this case we would be aiming for:

```
this collection and the list it contained on 
<b>
  {humanDate(since)}
</b>.
```

I haven't had a chance to investigate why this is happening yet.

File: src/printer.js [hunk 2]
```diff
@@ -3431,7 +3431,16 @@ function printJSXChildren(path, options, print, jsxWhitespace) {
           if (beginSpace) {
             children.push(jsxWhitespace);
           } else {
-            children.push(softline);
+            // Ideally this would be a `softline` to allow a break between
+            // tags and text.
+            // Unfortunately Facebook have a custom translation pipeline
+            // (https://github.com/prettier/prettier/issues/1581#issuecomment-300975032)
+            // that uses the JSX syntax, but does not follow the JSX whitespace
```
Comment (lydell) [inline src/printer.js]: I would call this "the React whitespace rules" since JSX does not seem to have whitespace rules.

**Comment by vjeux** [comment]
Thanks so much for working on this!

Could you also ensure that it works before as well?

```js
<div>Elements Elements Elements Elements Elements Elements Elements Elements (<span>{count}</span>) Elements Elements Elements</div>
```

currently outputs

```js
<div>
  Elements Elements Elements Elements Elements Elements Elements Elements (
  <span>{count}</span>) Elements Elements Elements
</div>;
```

which would add a space between `(` and `count`.

**Comment by karl** [comment]
I've pushed a quick change which hugs text to the beginning of tags as well.

Although there are some degenerative cases where you can end up with a line that never gets split.

Input:

```
this_really_should_split_across_lines =
  <div>
    before{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after
  </div>
```

Output:

```
this_really_should_split_across_lines = (
  <div>   before{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after
  </div>
);
```

I'll investigate what we might be able to do to fix this.

**Comment by vjeux** [comment]
Do people write this in practice?

**Comment by lydell** [comment]
@karl There's no way to split your last example, is it? Even if there is, I agree with @vjeux that it looks like a case where it's not worth spending time.

**Comment by karl** [comment]
Yep, totally agree that the degenerative case shouldn't block merging this!

**Comment by vjeux** [comment]
@lydell mind reviewing this PR?

**Comment by karl** [comment]
@lydell I believe it should be theoretically possible to split it by doing something like:

```
  <div>
    before{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{
      stuff
    }after{stuff}after{stuff}after{stuff}after{stuff}after{stuff}after{
      stuff
    }after{stuff}after{stuff}after
  </div>
```

File: tests/jsx-text-wrap/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (lydell) [inline tests/jsx-text-wrap/__snapshots__/jsfmt.spec.js.snap]: This PR looks great to me. Do you feel this is a blocker? I don't. We could always improve things in another PR.

Btw, I this would be a pretty nice formatting:

```
this collection and the list it contained on 
{" "}<b>{humanDate(since)}</b>.
```

Comment (vjeux) [inline tests/jsx-text-wrap/__snapshots__/jsfmt.spec.js.snap]: Nope, I don't really care about this.

**Comment by vjeux** [comment]
Just found a case where it's not stable the first time if you are interested. I don't think it's a big deal and it's okay if it's not fixed, but you may be interested.

```js
<div className="yourScore">
  Your score: <span className="score">{`${mini.crosstable.users[sessionUserId]} - ${mini.crosstable.users[user.id]}`}</span>
</div>
```

```js
<div className="yourScore">
  Your score:
  {" "}
  <span className="score">{`${mini.crosstable.users[sessionUserId]} - ${mini
    .crosstable.users[user.id]}`}</span>
</div>;
```

```js
<div className="yourScore">
  Your score:
  {" "}
  <span className="score">{`${mini.crosstable.users[sessionUserId]} - ${mini.crosstable.users[user.id]}`}</span>
</div>;
```

**Comment by karl** [comment]
@vjeux I've created an issue for the unstable formatting and added it to my TODO list.
